identity/aziot-identityd/src/identity.rs

```diff
@@ -719,6 +719,14 @@ impl IdentityManager {
         assert!(!status.eq_ignore_ascii_case("assigning"));
 
         let mut state = operation.registration_state.ok_or(Error::DeviceNotFound)?;
+
+        if state.status == Some("failed".to_string()) {
```

Comment:
[1.1's code](https://github.com/Azure/iotedge/blob/release/1.1/edgelet/dps/src/registration.rs#L188) seems to ignore this field and just assume that the presence of `.registration_state` means registration succeeded. In this code that would be handled by line 721 above. Are you seeing a case where registration fails but `registration_state` is still set to a valid device registration result?

Reply:
Yes, I was testing something else when I noticed if you change the IoT Hub access key to something invalid, you'll get a valid DeviceRegistrationResult:

```
DeviceRegistrationResult {
    tpm: None,
    x509: None,
    symmetric_key: None,
    registration_id: Some(
        "gowa-virtualbox-vm-1",
    ),
    created_date_time_utc: Some(
        "2021-05-25T17:50:50.1952288Z",
    ),
    assigned_hub: None,
    device_id: None,
    status: Some(
        "failed",
    ),
    substatus: None,
    error_code: Some(
        401002,
    ),
    error_message: Some(
        "{\"Message\":\"{\\\"errorCode\\\":401002,\\\"trackingId\\\":\\\"24821dbe7957463a853eefa84f067e0d-G:2-TimeStamp:05/25/2021 17:50:50\\\",\\\"message\\\":\\\"Unauthorized\\\",\\\"timestampUtc\\\":\\\"2021-05-25T17:50:50.1928721Z\\\"}\",\"ExceptionMessage\":\"\"}",
    ),
    last_updated_date_time_utc: Some(
        "2021-05-25T17:50:50.5209447Z",
    ),
    etag: Some(
        "Ijc0MDUyNTZhLTAwMDAtMDgwMC0wMDAwLTYwYWQzOGZhMDAwMCI=",
    ),
}
```

Reply:
Okay. So given that `assigned_hub` and `device_id` are still `None`, I would've thought provisioning would fail for *that* reason. But I see that lines 727-728 default to `""` in that case. Is that supposed to be done?

Reply:
Yes, provisioning will fail later on because of the device and hub being None.

But you'll get some misleading log messages in the meantime. Because this function thinks it succeeded, it'll print "Successfully reprovisioned." when it really hasn't. So I think it's better to catch the provisioning error here.

I'll change it to not default to "" though. It probably doesn't make sense to continue without knowing the device and hub names.

Reply:
>I'll change it to not default to "" though. It probably doesn't make sense to continue without knowing the device and hub names.

Yes, that's my point. If they're `.ok_or(...)?` instead then you still get the desired outcome of having the function fail, and it'd be "more correct" because those fields shouldn't be `null` in any case.

Reply:
Fixed